### DISCUSSION TITLE
How to do (specific kind of) interactive storytelling in Quasar?

### DISCUSSION THREAD

**Discussion by ropg** [quasarframework/quasar#11750] [General - Components / Directives / etc] [answered]
As an experiment, I wrote an interactive story environment, form scratch, in javascript, with no frameworks. You can see it in action in a short demo [here](https://polydemo.polyplot.de/) or with a full novel (in German) [here](gm.polyplot.de). I'd like to create a more mature version of this and am considering using Quasar for that.

In the continuum between games and literature, this project aims to stay closer to literature. One of the things we'd like to do is have the entire text be one continuous scroll: no clicking between sections. Right now this is achieved by simply adding more HTML to the DOM in the browser as the user click through the story, and that works remarkable well even for a half-megabyte story.

For the next version, I'd love to have the story be just text and Quasar components. However: ideally I must be able to make them appear in arbitrary order.

I've read through the entire documentation and while there's all sorts of dynamic loading of data, I don't see how to load new components themselves on a page dynamically. How would I go about this?

Any help would be greatly appreciated.

**Comment by pdanpdan** [comment, marked as answer]
The simplest solution would be to have the content in an array and render it with a v-for

**Comment by ropg** [comment]
Thank you for thinking about this!

Not sure that solves my problem. Suppose one of chapter 2A and 2B has to be rendered, based on a choice the reader makes. That can be a v-if, naturally. But then chapters 3 and 4 can appear in that order or reversed order, and they can in turn each contain different Quasar components.

So I guess one way framing my question is: can I dynamically change the order of components on a page?

Or, more generally, have my code construct an entirely new page at runtime ?

**Reply by rstoenescu** [reply]
Change the order in the array.

**Comment by ropg** [comment]
Forgive me for not getting something here, I'm sure I'm missing something. Suppose I modify the default Quasar App and make three minimal components like:

```vue
<template>
<q-card>
    <div class="text-h6">Number One</div>
</q-card>
</template>

<script>
export default { name: 'One' }
</script>
```

And then as Index.Vue, I have:

```vue
<template>
  <div>

    <q-btn @click="myArray.push('<One />')">One</q-btn>
    <q-btn @click="myArray.push('<Two />')">Two</q-btn>
    <q-btn @click="myArray.push('<Three />')">Three</q-btn>

    <One v-if="false" />
    <Two v-if="false" />
    <Three v-if="false" />

    <div v-for="(element) in myArray" :key="element">
        {{ element }}
    </div>

  </div>
</template>

<script>
import { defineAsyncComponent } from 'vue';
export default {
  name: "PageIndex",
  data () {
    return {
        myArray: [],
    }
  },
  components: {
    One: defineAsyncComponent(() => import('components/One.vue')),
    Two: defineAsyncComponent(() => import('components/Two.vue')),
    Three: defineAsyncComponent(() => import('components/Three.vue')),
  }
}
</script>
```

(Note that I include these components as never rendering because otherwise it complains about components defined but not used.) 

Then I get my three buttons:

![Screen Shot 2021-12-23 at 12 10 06](https://user-images.githubusercontent.com/343196/147232294-811646d6-6f03-4102-9732-5f2988d7fa18.png)

But when I press them I do not get my components rendered in arbitrary order, but just the literal text of what I add to the array:

![Screen Shot 2021-12-23 at 12 13 27](https://user-images.githubusercontent.com/343196/147232597-6f0a3d27-5dc7-44c5-b8ca-99cfdf296ebb.png)

How would I get it to render the components themselves in such a scenario?

Thank you for helping!

**Comment by pdanpdan** [comment]
Make a codepen

**Reply by ropg** [reply]
Solved it already, right after posting the question. Just didn't search for the right thing before. Thanks for replying..

**Comment by ropg** [comment]
Ha, I got the right Google Mojo and found Vue's "Dynamic Components" in this [forum post](https://forum.quasar-framework.org/topic/956/how-to-inject-an-element/2). For reference if anyone else needs the same thing, the above problem is solved with:

```vue
<template>
  <div>

    <q-btn @click="myArray.push('One')">One</q-btn>
    <q-btn @click="myArray.push('Two')">Two</q-btn>
    <q-btn @click="myArray.push('Three')">Three</q-btn>

    <One v-if="false" />
    <Two v-if="false" />
    <Three v-if="false" />

    <template v-for="(element) in myArray"  :key="element">
        <component :is="element"></component>
    </template>

  </div>
</template>

<script>
import { defineAsyncComponent } from 'vue';
export default {
  name: "PageIndex",
  data () {
    return {
        myArray: [],
    }
  },
  components: {
    One: defineAsyncComponent(() => import('components/One.vue')),
    Two: defineAsyncComponent(() => import('components/Two.vue')),
    Three: defineAsyncComponent(() => import('components/Three.vue')),
  }
}
</script>
```
